### PR TITLE
feat(findings): add evidence links

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- persist first-class finding evidence records that link findings and evaluation runs to claims, events, and graph roots
- expose bootstrap HTTP and ConnectRPC read APIs for listing and fetching finding evidence
- return evidence objects directly from runtime finding evaluation responses

## Testing
- make verify

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed conflict-resolution fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

File: internal/findings/service.go [hunk 1]
```diff
@@ -349,3 +661,81 @@ func findingIDs(findings []*ports.FindingRecord) []string {
...
+		seen[trimmed] = struct{}{}
+		unique = append(unique, trimmed)
+	}
+	sort.Strings(unique)
+	return unique
+}
+
+func findingEvidenceID(runtimeID string, findingID string, runID string) string {
+	replacer := strings.NewReplacer(" ", "-", "_", "-", "/", "-", ":", "-", ".", "-")
+	prefix := replacer.Replace(strings.TrimSpace(runtimeID) + "-" + strings.TrimSpace(findingID) + "-" + strings.TrimSpace(runID))
```
Comment (josh-writer) [inline internal/findings/service.go]: NIT/edge case Evidence IDs can collide after lossy normalization. findingEvidenceID maps _, /, :, ., and spaces to -, so distinct finding IDs like finding-a_b and finding-a-b in the same runtime/run produce the same evidence ID; the second upsert overwrites the first evidence record

File: internal/bootstrap/app.go [hunk 2]
```diff
@@ -658,19 +778,57 @@ func (s *bootstrapService) GetFindingEvaluationRun(ctx context.Context, req *con
...
+	evidence, err := findings.New(
+		sourceRuntimeStore(s.deps.StateStore),
+		eventReplayer(s.deps.AppendLog),
+		findingStore(s.deps.StateStore),
+		findingEvaluationRunStore(s.deps.StateStore),
+		findingEvidenceStore(s.deps.StateStore),
+		claimStore(s.deps.StateStore),
+	).GetEvidence(ctx, req.Msg.GetId())
+	if err != nil {
+		return nil, err
```
Comment (josh-writer) [inline internal/bootstrap/app.go]: GetFindingEvidence returns missing evidence as a plain error.  When GetEvidence returns ports.ErrFindingEvidenceNotFound, connect-go will encode it as CodeUnknown/HTTP 500  instead of NotFound, unlike the HTTP handler’s 404 mappin

**Review by josh-writer — CHANGES_REQUESTED** [review]

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-reviewed the live unresolved review threads for this PR (including outdated ones) against the current branch tip, cross-checking the latest parent-session snapshot as requested.

Checked/confirmed addressed:
- failed-run `EventsEvaluated` accounting in `EvaluateSourceRuntime`
- evidence deduplication in `EvaluateSourceRuntime`
- stable returned evidence `CreatedAt` via persisted evidence reload
- failed-run `EventsEvaluated` accounting in `EvaluateSourceRuntimeRules`

No code changes were needed on the current head, so no commit/push was required.

Validation run:
- `go test ./internal/findings -run 'TestEvaluateSourceRuntimeFindings(DeduplicatesEvidenceByID|FailedRunCountsOnlyEvaluatedEvents)|TestEvaluateSourceRuntimeRules(FailedRunCountsOnlyEvaluatedEvents|DeduplicatesEvidenceByID)'`\n- `make verify`\n\nRemaining unresolved review threads after this audit: none.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review this evidence-links PR deeply. Areas I want you to stress:

1. **Tenant/runtime scoping**: evidence links must never let a finding in one tenant/runtime reference, reveal, overwrite, or delete evidence owned by another tenant/runtime.
2. **Upsert/idempotency semantics**: check conflict keys, uniqueness assumptions, replacement behavior, and whether repeated writes or concurrent writes can duplicate, orphan, or silently mutate links.
3. **Authorization and data exposure**: make sure read APIs only expose links for findings the caller is allowed to see, and that evidence URNs/metadata do not leak hidden graph/report/source details.
4. **Lifecycle correctness**: review deletion/retraction paths, finding replacement, replay/rebuild behavior, migrations, and whether stale evidence remains after a finding changes fingerprint/id/status.
5. **Pagination and ordering**: look for unstable ordering, missing limits, cursor bugs, or accidental full-table scans when listing links for high-cardinality findings.
6. **Contract compatibility**: verify protobuf/JSON/API shape and SDK expectations are consistent with adjacent finding-read/evaluation-run PRs.

**Comment by jonathanhaaswriter** [comment]
Closing this PR — its substantive feature work is already in `main` via the earlier consolidation rollups, and the remaining unresolved factory-droid findings on this stack have been addressed in main via the follow-up PR:

- #500 (`fix(findings): persist normalized event limit and disambiguate run/evidence ids`)

PR #500 specifically pulls forward:

- Persisting the **normalized** event limit on `FindingEvaluationRun` records (PR #477)
- Hashing raw runtime/rule ids and appending a crypto-rand suffix to `findingEvaluationRunID` so concurrent multi-rule evaluations and normalization-equivalent ids never collide (PR #479)
- Appending a sha256 suffix to `findingEvidenceID` so lossy normalization no longer overwrites distinct evidence rows (PR #479)

This branch has diverged thousands of lines from `main` and rebasing it cleanly is no longer practical; closing here keeps the open queue accurate. If anything from this stack still needs to land, please open a fresh PR cut from `main`.
